### PR TITLE
ATHATH finally goes off the deep end and touches   L U N G   C O D E

### PR DESCRIPTION
**Pull request by ATH1909**
## About The Pull Request

https://www.youtube.com/watch?v=tCpMIe8FjEU

glubtok discovered that the main intended effect of plasma fixation (standing in plasma gas heals you) didn't actually work in this issue report: https://github.com/tgstation/tgstation/issues/46055. He posted a fix for the issue in his issue report, which I have implemented in this PR. He also lamented in said issue report that plasma fixation doesn't have any interactions with you BREATHING plasma, which led me to try to figure out a way to add said interactions.

The result of my efforts is this PR.

I've added two new variables to lungs. One is the all_you_need_is_plasma variable, which (basically) lets plasma count as other gases for the purposes of determining if you're suffocating from a lack of a certain gas (just setting safe_toxins_max to 0 won't work, since humans and such will still suffocate from a lack of oxygen while breathing a pure plasma mixture). The other variable is the plasma_healing variable, which causes each breath of (enough) plasma you take to heal you for one point in each of the four primary damage types (I would like to point out that you usually only breathe once per 4 ticks; in other words, this regen should be at around 1/4th the speed of the regen that podpeople in light get). The toxin healing part of this healing will heal even toxinlovers, such as jellypeople These variables have been set to TRUE for vacuoles (a.k.a. slime lungs), which means that jellypeople can now breathe plasma without suffocating and can heal while breathing plasma (it should be noted that safe_toxins_max had already been set to 0 by default for their lungs before I made this PR). Because they can now breathe plasma far more safely, the blood regeneration rate that jellypeople receive from breathing plasma has been halved in speed.

The plasma fixation virus symptom now has two additional threshold effects. The new resistance 4 threshold effect permanently sets the safe_toxins_max and all_you_need_is_plasma variables of the lungs of hosts of its virus to 0 and TRUE, respectively. The new resistance 10 threshold effect permanently sets the plasma_healing variable of the lungs of hosts of the virus to TRUE.

Also, because I was modifying heal.dmm anyway, I gave the regenerative coma symptom the ability to give hosts of its virus the TRAIT_NOCRITDAMAGE trait, which should halt the natural oxygen (or brute, if you have the TRAIT_NOBREATH trait or the like) damage from being in crit while you have the virus (unless you're a monkey or the like, I think, because non-human carbons have weird crit damage code (one part of it even directly checks for the presence epinephrine chem!)).

## Why It's Good For The Game

Plasma fixation's main effect literally doesn't work right now. The new threshold effects added to plasma fixation encourage creative usages of atmos and/or the purchase and distribution of plasmaman supply kits from Cargo in order to give the crew a mild healing effect. Plus, the "ability to breathe plasma" one is quite flavorful/cool (IMO).

## Changelog
:cl: ATHATH
add: The plasma fixation virus symptom now has two additional threshold effects. The new resistance 4 threshold effect gives you the ability to safely breathe even pure plasma mixtures, while the new resistance 10 threshold effect heals you each time you breathe in (a significant amount of) plasma.
fix: The plasma fixation virus symptom now actually heals you while you are standing in plasma gas. Credit goes to glubtok for pointing this bug out and writing the code for this fix for it.
balance: The regenerative coma symptom now stabilizes you while you're in critical condition.
tweak: Vacuoles (a.k.a. slime lungs (the lungs of jellypeople and their ilk) now allow you to safely breathe even pure plasma mixtures (instead of suffocating you due to a lack of oxygen) with them and heal whenever you breathe in a significant amount of plasma with them.
balance: Since pure plasma mixtures are no longer as harmful to jellypeople as they were before, the blood gain rate from breathing plasma with a vacuole has been halved.
/:cl:

### CODE REVIEW

**Comment by ExcessiveUseOfCobblestone** [comment]
I don't want perma effects coming out of viro, as we already have a mechanism by which people can upgrade their lungs (cybernetics).

**Comment by Whoneedspacee** [comment]
lung code bad pr good

**Comment by Ratvar** [comment]
Cybernetics and surgeries AND synth stuff suck in general. You gotta 
0. Have synth organs be researched by R&D instead of them powergaming with nanites 
1. Find someone competent to do surgery 
2. Have that someone be on your side 
3. Have that someone be willing to waste their time 
4. Waste your own time doing surgery 
5. Not be attacked in the mean time 
6. Finally get your lungs and instadie if there's any antag+officers with EMP, which is everyone but revs (iron uranium from protolathe, anyone?) due to borgs being so OP

Having to breath plasma means that you cannot just spread it to everyone, that you'll need to prepare air supply, that you'll need to wear a mask all the time... which is 10000% times better then time waste and EMP weakness of surgery that's never used because it's so bad and also worse then getting some chemicals to heal yourself beforehand. Cybernetics suck, rework EMP weakness and surgeries in general, this PR is good and isn't just old brainless sneeze baldness fleshmending viro combo

**Comment by ATH1909** [comment]
> I don't want perma effects coming out of viro, as we already have a mechanism by which people can upgrade their lungs (cybernetics).

The hyphema, deafness, and deoxyrybonucleic saboteur symptoms (and maybe the disfiguration and/or skin pigmentation symptoms) can all have permanent effects (even after the removal of the virus); why can't this symptom?

Also, it'd be really, really hard to make the effects get reverted upon the removal of the disease without opening up a few "gaps" in the coverage of the reversion effect. I could try to reset the variables of the modified lungs to what they were before when the disease is cured, but what happens if the lungs are surgically removed before the disease is cured (and then placed back in the body after the disease is cured)? Basically, I can't stop people from keeping the effects of this symptom after its virus is cured, so I might as well make it a feature.

**Comment by Ratvar** [comment]
I wouldn't say normal people would cut out their lungs once they get a good virus just so they can stack it on top of whatever other healing virus they have. With your feature, people'll just always have Plasma Fixation an extra unpurgeable virus symptom. Could as well slap bugabusers with mop instead of just sighing and saying "fine".

**Comment by ATH1909** [comment]
> I wouldn't say normal people would cut out their lungs once they get a good virus just so they can stack it on top of whatever other healing virus they have. With your feature, people'll just always have Plasma Fixation an extra unpurgeable virus symptom. Could as well slap bugabusers with mop instead of just sighing and saying "fine".

I added an extra paragraph to my post just before (or was it after?) you posted your reply post; you might want to take a look at it.

**Comment by ATH1909** [comment]
Also, people suddenly starting to suffocate from/get poisoned by plasma after being infected with some random miasma disease (or a sentient disease) seems kind of lame.

**Comment by Ratvar** [comment]
You could try making a virus that's actually resistant to being overtaken and wash themselves. Them breathing from tank of plasma counteracts miasma.

**Comment by MCterra10** [comment]
That's a lot of ~~damage~~ commits

**Comment by ATH1909** [comment]
> That's a lot of ~damage~ commits

Yeah, I recently deleted my entire fork (instead of just my master branch like I was supposed to), so those should be gone in my future PRs.

**Comment by XDTM** [comment]
a lung variable dedicated to a single symptom doesn't seem like a good idea to me

the proper way to achieve this without snowflake code would be to send a signal on breathe with the gasmix and add a listener to the symptom

**Comment by Akrilla** [comment]
madman straight up webeditting

**Comment by nemvar** [comment]
The proper fix would be to remove all breathing code and write it all again. Imo.
```
/mob/living/carbon/handle_breathing(times_fired)
	var/next_breath = 4
	var/obj/item/organ/lungs/L = getorganslot(ORGAN_SLOT_LUNGS)
	var/obj/item/organ/heart/H = getorganslot(ORGAN_SLOT_HEART)
	if(L)
		if(L.damage > L.high_threshold)
			next_breath--
	if(H)
		if(H.damage > H.high_threshold)
			next_breath--

	if((times_fired % next_breath) == 0 || failed_last_breath)
		breathe() //Breathe per 4 ticks if healthy, down to 2 if our lungs or heart are damaged, unless suffocating
		if(failed_last_breath)
			SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "suffocation", /datum/mood_event/suffocation)
		else
			SEND_SIGNAL(src, COMSIG_CLEAR_MOOD_EVENT, "suffocation")
	else
		if(istype(loc, /obj/))
			var/obj/location_as_object = loc
			location_as_object.handle_internal_lifeform(src,0)
```
This is just the first of like 4 giant undocumented and confusing procs (and almost every single one does ``getorganslot(ORGAN_SLOT_LUNGS)``) that breathing goes through. This thing alone handles like 5 different things and has confusing and bad math. breathe() gets overridden in human and then handled in species so it doesn't even have shared functionality.

Basically, what I'm saying is that this isn't much of an improvement. It makes an existing problem (bad convoluted code) worse and the gain in functionality is also questionable.

Want my recommendation? Either don't touch this at all or just replace the magic strings with the plasma typepath and be done with it.

**Comment by ExcessiveUseOfCobblestone** [comment]
> The hyphema, deafness, and deoxyrybonucleic saboteur symptoms (and maybe the disfiguration and/or skin pigmentation symptoms) can all have permanent effects (even after the removal of the virus); why can't this symptom?

Outside of pigmentation which is cosmetic, most if not all of these can be fixed with Chem. Push comes to shove you have to replace via surgery.

Perhaps a more obvious comparison though is that these are all BAD and you WANT to get these fixed. Someone looking to take away your lung perks is going to have a much harder time than someone looking to take away your eye/ear/genetic debuffs.

**Comment by ATH1909** [comment]
Ah shit, I forgot that you breathe twice as often while your lungs are past the high damage threshold. That means that I'm gonna have to halve the healing rate from plasma_healing for severely damaged lungs.



> a lung variable dedicated to a single symptom doesn't seem like a good idea to me
> 
> the proper way to achieve this without snowflake code would be to send a signal on breathe with the gasmix and add a listener to the symptom

And how would a signal allow people to breathe pure plasma? Just setting safe_toxins_max to 0 won't work for that.

**Comment by ATH1909** [comment]
> > The hyphema, deafness, and deoxyrybonucleic saboteur symptoms (and maybe the disfiguration and/or skin pigmentation symptoms) can all have permanent effects (even after the removal of the virus); why can't this symptom?
> 
> Outside of pigmentation which is cosmetic, most if not all of these can be fixed with Chem. Push comes to shove you have to replace via surgery.
> 
> Perhaps a more obvious comparison though is that these are all BAD and you WANT to get these fixed. Someone looking to take away your lung perks is going to have a much harder time than someone looking to take away your eye/ear/genetic debuffs.

I could add a bit in on_remove() (or whatever proc organs call when they are removed) for lungs that causes their safe_toxins_max, all_you_need_is_plasma, and plasma_healing variables to revert back to their initial states, then make the plasma fixation symptom do a similar thing to that when its virus is cured. Would that be acceptable?

**Comment by ATH1909** [comment]
Also, I'd like to mention that I'm probably going to have to close this PR and open a new one that's near-identical to it so that I can edit it, since I messed up and deleted my entire fork a few days ago (which means that I can't edit this PR anymore).

**Comment by Cyprex** [comment]
> Someone looking to take away your lung perks is going to have a much harder time than someone looking to take away your eye/ear/genetic debuffs.

Why is the difficulty of taking away someones perk even relevant? Also, just take away their internals?

**Comment by ExcessiveUseOfCobblestone** [comment]
> 
> 
> > Someone looking to take away your lung perks is going to have a much harder time than someone looking to take away your eye/ear/genetic debuffs.
> 
> Why is the difficulty of taking away someones perk even relevant? Also, just take away their internals?

Because when you add something that potentially makes it harder to kill people in certain ways there naturally needs to be a way to mitigate that? 

Making it easier to keep people alive is just as important as features that make it easier to kill someone...

**Comment by ExcessiveUseOfCobblestone** [comment]
> I could add a bit in on_remove() (or whatever proc organs call when they are removed) for lungs that causes their safe_toxins_max, all_you_need_is_plasma, and plasma_healing variables to revert back to their initial states, then make the plasma fixation symptom do a similar thing to that when a virus that has it is cured. Would that be acceptable?

As long as there are no permanent benefits to having a (theoretically) temporary beneficial virus, I won't have an issue with it conceptually, I didn't even bother trying to use my powergamer excel with this yet to see what kinda stuff I can make:

**Comment by Cyprex** [comment]
Excuse me, we let people powergame with nanites which are arguably harder to deal with yet healing via a special internals tank after needing to already make a virus is too much?

**Comment by MCterra10** [comment]
I guess this symptom will remain entirely useless and a waste of space for a while longer yet
